### ISSUE TITLE
Lock the dependencies when building

### ISSUE THREAD

**Issue by ethanfrey** [confio/cosmwasm-opt#5]
`cargo build` (and `wasm-pack build`) will always use the deps referenced in `Cargo.toml`, but may update `Cargo.lock` with eg. patch releases. In some cases, this can change the actual code in the build path and thus the wasm bytecode/hash. We want to avoid this, so a contract can be safely compiled 6 months later with the same hash.

We should fix the line https://github.com/confio/cosmwasm-opt/blob/master/optimize.sh#L8 to pass in a flag to `cargo`. I think adding `-- --locked` will do this, but need to verify

**Comment by ethanfrey** [comment]
I tried this out with the escrow contract on `escrow-0.2.0` branch on cosmwasm-examples. It has `cosmwasm="0.6.0"` in `Cargo.lock`, and `cosmwasm="~0.6.0"` in  `Cargo.toml`. At the time of the experiment, `0.6.3` was available.

* If I ran (with a previous cache), it builds the same as before.
* If I update the `Cargo.toml` to `cosmwasm="0.6.3"`, it will update the `Cargo.lock` file and produce a build with a different hash (naturally)
* If I remove the docker cache (`docker volume rm escrow_cache`), and rebuild, it still stays the same.

In other words, as long as `Cargo.toml` is compatible with `Cargo.lock`, it will not update anything.
If they are incompatible, it will update `Cargo.lock` as well. Thus, the problem is not big, but I can still add the flag and check for errors in this case

**Comment by ethanfrey** [comment]
Weird behavior trying `--locked`.

Get code locked to older version:

```sh
git clone https://github.com/confio/cosmwasm-examples
cd cosmwasm-examples
git checkout escrow-0.2.0
cd escrow
```

Update the version in `Cargo.toml` only: 

`sed -i 's/0.6.0/0.6.3/' Cargo.toml`

View deps before:

```
$ grep cosmwasm Cargo.*

Cargo.lock:name = "cosmwasm"
Cargo.lock:name = "cosmwasm-vm"
Cargo.lock: "cosmwasm 0.6.0 (registry+https://github.com/rust-lang/crates.io-index)",
Cargo.lock: "cosmwasm 0.6.0 (registry+https://github.com/rust-lang/crates.io-index)",
Cargo.lock: "cosmwasm-vm 0.6.0 (registry+https://github.com/rust-lang/crates.io-index)",
Cargo.lock:"checksum cosmwasm 0.6.0 (registry+https://github.com/rust-lang/crates.io-index)" = "62d3a410d13a98b07c14f8be00656f24349010951a8f276f87bec5cb58221331"
Cargo.lock:"checksum cosmwasm-vm 0.6.0 (registry+https://github.com/rust-lang/crates.io-index)" = "1134b57ca4b5956331c089253616a4ed98fc6e8822a567b0f7cb89a311191227"

Cargo.toml:repository = "https://github.com/confio/cosmwasm-examples"
Cargo.toml:backtraces = [ "cosmwasm/backtraces", "cosmwasm-vm/backtraces" ]
Cargo.toml:cranelift = [ "cosmwasm-vm/default-cranelift"]
Cargo.toml:singlepass = [ "cosmwasm-vm/default-singlepass"]
Cargo.toml:cosmwasm = { version = "~0.6.3" }
Cargo.toml:cosmwasm-vm = { version = "~0.6.3", default-features = false }
```

Build it:

```
cargo build --locked
```

This should error, and does sometimes...
